doc/man/mubdi-cli.1
.\" DO NOT MODIFY THIS FILE!  It was generated by help2man 1.47.6.
.TH MuBdI-CLI "1" "August 2021" "mubdi-cli v6.7.6" "User Commands"
.SH NAME
mubdi-cli \- manual page for mubdi-cli v6.7.6
.SH SYNOPSIS
.B mubdi-cli
[\fI\,options\/\fR] \fI\,<command> \/\fR[\fI\,params\/\fR]  \fI\,Send command to MuBdI Core\/\fR
.br
.B mubdi-cli
[\fI\,options\/\fR] \fI\,-named <command> \/\fR[\fI\,name=value\/\fR]... \fI\,Send command to MuBdI Core (with named arguments)\/\fR
.br
.B mubdi-cli
[\fI\,options\/\fR] \fI\,help                List commands\/\fR
.br
.B mubdi-cli
[\fI\,options\/\fR] \fI\,help <command>      Get help for a command\/\fR
.SH DESCRIPTION
MuBdI Core RPC client version v6.7.6\-charlotte
.SH OPTIONS
.HP
\-?
.IP
This help message
.HP
\fB\-conf=\fR<file>
.IP
Specify configuration file (default: mubdi.conf)
.HP
\fB\-datadir=\fR<dir>
.IP
Specify data directory
.PP
Chain selection options:
.HP
\fB\-testnet\fR
.IP
Use the test chain
.HP
\fB\-regtest\fR
.IP
Enter regression test mode, which uses a special chain in which blocks
can be solved instantly. This is intended for regression testing tools
and app development.
.HP
\fB\-named\fR
.IP
Pass named instead of positional arguments (default: false)
.HP
\fB\-rpcconnect=\fR<ip>
.IP
Send commands to node running on <ip> (default: 127.0.0.1)
.HP
\fB\-rpcport=\fR<port>
.IP
Listen for JSON\-RPC connections on <port> (default: 51473 or testnet:
51475)
.HP
\fB\-rpcwait\fR
.IP
Wait for RPC server to start
.HP
\fB\-rpcuser=\fR<user>
.IP
Username for JSON\-RPC connections
.HP
\fB\-rpcpassword=\fR<pw>
.IP
Password for JSON\-RPC connections
.HP
\fB\-rpcclienttimeout=\fR<n>
.IP
Timeout in seconds during HTTP requests, or 0 for no timeout. (default:
900)
.HP
\fB\-rpcwallet=\fR<walletname>
.IP
Send RPC for non\-default wallet on RPC server (needs to exactly match
corresponding \fB\-wallet\fR option passed to mubdid)
.SH COPYRIGHT
Copyright (C) 2009-2021 The Bitcoin Core Developers

Copyright (C) 2014-2021 The Dash Core Developers

Copyright (C) 2015-2021 The PIVX Core Developers

Copyright (C) 2010-2022 Nur1Labs Developers

This is experimental software.

Distributed under the MIT software license, see the accompanying file COPYING
or <http://www.opensource.org/licenses/mit-license.php>.

This product includes UPnP software written by Thomas Bernard.


ReleaseNotes.md
Before 3.0 is 2.9 We notes it's moved from Github Repository and removed there because conflicted of public commit. 

Year 2015-2017

Release notes 3.0

- upload to github
- added change for files
- reset work

Release notes 3.0.1

- change cores
- update setting
- update rpc and blockindicies
- added checkpoint
  a. checkpoint cores
  b. checkpoint payment_gate

Release notes 3.0.2

- checpoint added
  - recover lost checkpoint
  - updates new chain run and simple
  - fixed chain
- checkpoint 6/4/2018
  - improve chain
  - wallet log
- checkpoint 6/5/2018
  - added log
  - improve wallet
  - remove unecessary file

Release notes 3.0.3

- checkpoint 6/6/2018
  - update transfer
  - fixed update for autoremove transaction chain
  - faster block

- checkpoint 6/9/2018
  - improve wallet from bytecoin models
  - crypto patch from karbo
  - upgrade core system with PoS and MN support

Release notes 3.0.4

- checkpoint 6/14/2018
  - improve hash cores
  - improve masternodes
  - improve PoS
  - super nodes increase rpc value
  - remotes checkpoint
  - dust test transaction
  - long hash stop for solo on SN[Super Nodes][like gpu solos]
  - block point added
  - parent block fixed
  - added confirmation log rpc
  - security increases
  - android support

- checkpoint 6/15/2018
  - revert work to checkpoint date 13
  - ignore aliases bug
  - ignore passwork on simple_wallet cuz bug on address
  - made checkpoint today 12:00 pm

- checkpoint 6/16/2018
  - fixed alias as bug. removed!
  - revert to old checkpoint. upgrade checkpoint level
  - new string on daemon
  - android fixed

- checkpoint 6/17/2018
  - cross chain added
  - fixed old wallet bug
  - backup fixed
  - fixed alias by move to android

Release notes 3.0.5

- checkpoint 6/19/2018
  - added merchant wallet[building]
  - added masternodes wallet[building]
  - fixed nice hash attack
  - alternative chain fixed

- checkpoint 6/20/2018
  - transaction id fixed
  - fixed bug from loop diff

- checkpoint 6/21/2018
  - fixed pool transaction
  - fixed diff
  - fixed merge mining
  - wallet sync fixed
  - revert back merge mining to currency

- checkpoint 6/22/2018
  - fixed mem-pool faster
  - added ban for mempool
  - added print ban

Release notes 3.0.6

- checkpoint 6/24/2018
  - implemeted LWMA-2 version fixed
  - upgrade to for new major block 5
  - fixed bug and remove LWMA-1 trace

- checkpoint 6/25/2018
  - tidy structure
  - added on transaction
  - added max mixin
  - fixed cores transaction
  - fixed fusion + fee transaction
  - block chain faster
  - add top block hash
  - fixed gui build
  - add lite wallet build
  - rename class to new version

- checkpoint 6/26/2018
  - add doxyfile by homdx
  - fixed capitalize 0 or o
  - welcome to homdx
  - added gui

- checkpoint 6/27/2018
  - added rpc auth
  - added block fee by niobio[disable and enable feature]
  - fixed some minor bug
  - pool size using command line
  - remove annoyed  tx verification failed messages
  - remove double name testnet prefix
  - Reduce log messages on pushing block

Release notes 3.0.7

- checkpoint 6/28/2018
  - implemented heavy
  - better pow call
  - password fixed
  - fixed slow hash
  - add The Niobio Developer as contibutor list
  - The Italo Developer as contibutor list
  - fixed major bug on heavy
  - get_payments method now returns confirmations

- checkpoint 6/30/2018
  - some rest and idle fixed

- checkpoint 7/1/2018
  - upgrading platform
  - code improved

Release notes 3.0.8

- checkpoint 7/2/2018
  - code upgrade
  - dynamic modular integrated
  - minor breaker
  - updates some string
  - updates blockchain rules
  - updates modular

- checkpoint 7/3/2018
  - enable slow reward and slow miner
  - fixed minor bugs

- checkpoint 7/4/2018
  - masternodes implemented
  - bug fixed major

- checkpoint 7/6/2018
  - fixed some bug masternodes

- checkpoint 7/7/2018
  - added validate address
  - added new cells

- checkpoint 7/9/2018
  - added split cell[4]

- checkpoint 7/16/2018
  - minor fixed

- checkpoint 7/17/2018
  - minor fixed

Release notes 3.0.9

- checkpoint 7/18/2018
  - updates core
  - change structure
  - fixed cells

Release notes 4.0.x

- checkpoint 7/28/2018
  - change platform
  - Enable point of sales
  - Enable Minting
  - Enable supernodes
  - Fee controls
  - New Platform
  - Dns Auto Fixed
  - Generate reward by random difficulty

- checkpoint 7/29/2018
  - fatal coin orphan fixed
  - difficulty control
  - tokens is not overloaded
  - minting check and control reward fixed
  
- checkpoint 8/3/2018
  - fixed genesis block
  - platform rebuilding
  - fixed seed
  - hardcode genesis
  - emercoin fork finish
  - emer seed remove

- checkpoint 8/4/2018
  - dns emer sync with labs
  - bitcoin code merge zEtA platform
  - consesus updates fork emercoin 8/3/2018
  - prepare zEtA files
  - consesus updates fork emercoin 8/4/2018

- checkpoint 8/8/2018
  - remove emercoin hash
  - branding
  - masternodes
  - fixed some brand
  - issue chain conflict dash fixed
  - updates bignum fixed by emercoin developer and peer developer

- checkpoint 8/11/2018
  - mining rpc fixed by emercoin
  - rpc reading stake by emercoin

- checkpoint 8/13/2018
  - updates gui final
  - fixed gui and remove prefix
  - tags remove updates code cores

Release notes 4.2.x

- checkpoint 8/15/2018
  - updates segwit 8
  - restructure MuBdI
  - cores code revert
  - gui fixed
  - MuBdI connector[struct]
  - BarkleyDB 6.x support

- checkpoint 8/20/2018
  - build 2 system integrated
  - fix rpcmining/getblocktemplate univalue transition logic error
  - add getfeeinfo RPC
  - [RPC] add setban/listbanned/clearbanned RPC Commands
  - add disconnectnode RPC command
  - introduce banlist.dat to store the ban list across restarts
  - introduce CScheduler lightweight task scheduler]
  - fix `importzercoins` for use with UniValue]
  - [RPC] implement random-cookie based authentication
  - findserial RPC
  - return change to sender when minting
  - little branding

src/mubdi-cli.cpp
/***********************************************************************
**************Copyright (c) 2009-2010 Satoshi Nakamoto******************
***********Copyright (c) 2009-2021 The Bitcoin developers***************
*************Copyright (c) 2009-2015 The Dash developers****************
*************Copyright (c) 2015-2021 The PIVX developers****************
******************Copyright (c) 2010-2023 Nur1Labs**********************
>Distributed under the MIT/X11 software license, see the accompanying
>file COPYING or http://www.opensource.org/licenses/mit-license.php.
************************************************************************/

#if defined(HAVE_CONFIG_H)
#include "config/mubdi-config.h"
#endif

#include "chainparamsbase.h"
#include "clientversion.h"
#include "fs.h"
#include "rpc/client.h"
#include "rpc/protocol.h"
#include "util/system.h"
#include "utilstrencodings.h"

#include <stdio.h>
#include <tuple>

#include <event2/buffer.h>
#include <event2/keyvalq_struct.h>
#include "support/events.h"

#include <univalue.h>

static const char DEFAULT_RPCCONNECT[] = "127.0.0.1";
static const bool DEFAULT_NAMED=false;
static const int DEFAULT_HTTP_CLIENT_TIMEOUT=900;

std::string HelpMessageCli()
{
    const auto defaultBaseParams = CreateBaseChainParams(CBaseChainParams::MAIN);
    const auto testnetBaseParams = CreateBaseChainParams(CBaseChainParams::TESTNET);
    std::string strUsage;
    strUsage += HelpMessageGroup("Options:");
    strUsage += HelpMessageOpt("-?", "This help message");
    strUsage += HelpMessageOpt("-conf=<file>", strprintf("Specify configuration file (default: %s)", MUBDI_CONF_FILENAME));
    strUsage += HelpMessageOpt("-datadir=<dir>", "Specify data directory");
    AppendParamsHelpMessages(strUsage);
    strUsage += HelpMessageOpt("-named", strprintf("Pass named instead of positional arguments (default: %s)", DEFAULT_NAMED));
    strUsage += HelpMessageOpt("-rpcconnect=<ip>", strprintf("Send commands to node running on <ip> (default: %s)", DEFAULT_RPCCONNECT));
    strUsage += HelpMessageOpt("-rpcport=<port>", strprintf("Listen for JSON-RPC connections on <port> (default: %u or testnet: %u)", defaultBaseParams->RPCPort(), testnetBaseParams->RPCPort()));
    strUsage += HelpMessageOpt("-rpcwait", "Wait for RPC server to start");
    strUsage += HelpMessageOpt("-rpcuser=<user>", "Username for JSON-RPC connections");
    strUsage += HelpMessageOpt("-rpcpassword=<pw>", "Password for JSON-RPC connections");
    strUsage += HelpMessageOpt("-rpcclienttimeout=<n>", strprintf("Timeout in seconds during HTTP requests, or 0 for no timeout. (default: %d)", DEFAULT_HTTP_CLIENT_TIMEOUT));
    strUsage += HelpMessageOpt("-rpcwallet=<walletname>", "Send RPC for non-default wallet on RPC server (needs to exactly match corresponding -wallet option passed to mubdid)");

    return strUsage;
}

/** libevent event log callback */
static void libevent_log_cb(int severity, const char *msg)
{
#ifndef EVENT_LOG_ERR // EVENT_LOG_ERR was added in 2.0.19; but before then _EVENT_LOG_ERR existed.
# define EVENT_LOG_ERR _EVENT_LOG_ERR
#endif
    // Ignore everything other than errors
    if (severity >= EVENT_LOG_ERR) {
        throw std::runtime_error(strprintf("libevent error: %s", msg));
    }
}

//////////////////////////////////////////////////////////////////////////////
//
// Start
//

//
// Exception thrown on connection error.  This error is used to determine
// when to wait if -rpcwait is given.
//
class CConnectionFailed : public std::runtime_error
{
public:
    explicit inline CConnectionFailed(const std::string& msg) : std::runtime_error(msg)
    {
    }
};

static bool AppInitRPC(int argc, char* argv[])
{
    //
    // Parameters
    //
    gArgs.ParseParameters(argc, argv);
    if (argc < 2 || gArgs.IsArgSet("-?") || gArgs.IsArgSet("-h") || gArgs.IsArgSet("-help") || gArgs.IsArgSet("-version")) {
        std::string strUsage = PACKAGE_NAME " RPC client version " + FormatFullVersion() + "\n";
        if (!gArgs.IsArgSet("-version")) {
            strUsage += "\n"
                        "Usage:  mubdi-cli [options] <command> [params]  Send command to " PACKAGE_NAME "\n"
                        "or:     mubdi-cli [options] -named <command> [name=value]... Send command to " PACKAGE_NAME " (with named arguments)\n"
                        "or:     mubdi-cli [options] help                List commands\n"
                        "or:     mubdi-cli [options] help <command>      Get help for a command\n";
            strUsage += "\n" + HelpMessageCli();
        }

        fprintf(stdout, "%s", strUsage.c_str());
        return false;
    }
    if (!fs::is_directory(GetDataDir(false))) {
        fprintf(stderr, "Error: Specified data directory \"%s\" does not exist.\n", gArgs.GetArg("-datadir", "").c_str());
        return false;
    }
    try {
        gArgs.ReadConfigFile(gArgs.GetArg("-conf", MUBDI_CONF_FILENAME));
    } catch (const std::exception& e) {
        fprintf(stderr, "Error reading configuration file: %s\n", e.what());
        return false;
    }
    // Check for -testnet or -regtest parameter (BaseParams() calls are only valid after this clause)
    try {
        SelectBaseParams(gArgs.GetChainName());
    } catch(const std::exception& e) {
        fprintf(stderr, "Error: %s\n", e.what());
        return false;
    }
    if (gArgs.GetBoolArg("-rpcssl", false))
    {
        fprintf(stderr, "Error: SSL mode for RPC (-rpcssl) is no longer supported.\n");
        return false;
    }
    return true;
}


/** Reply structure for request_done to fill in */
struct HTTPReply
{
    HTTPReply(): status(0), error(-1) {}

    int status;
    int error;
    std::string body;
};

const char *http_errorstring(int code)
{
    switch(code) {
        case EVREQ_HTTP_TIMEOUT:
            return "timeout reached";
        case EVREQ_HTTP_EOF:
            return "EOF reached";
        case EVREQ_HTTP_INVALID_HEADER:
            return "error while reading header, or invalid header";
        case EVREQ_HTTP_BUFFER_ERROR:
            return "error encountered while reading or writing";
        case EVREQ_HTTP_REQUEST_CANCEL:
            return "request was canceled";
        case EVREQ_HTTP_DATA_TOO_LONG:
            return "response body is larger than allowed";
        default:
            return "unknown";
    }
}

static void http_request_done(struct evhttp_request *req, void *ctx)
{
    HTTPReply *reply = static_cast<HTTPReply*>(ctx);

    if (req == NULL) {
        /* If req is NULL, it means an error occurred while connecting: the
         * error code will have been passed to http_error_cb.
         */
        reply->status = 0;
        return;
    }

    reply->status = evhttp_request_get_response_code(req);

    struct evbuffer *buf = evhttp_request_get_input_buffer(req);
    if (buf)
    {
        size_t size = evbuffer_get_length(buf);
        const char *data = (const char*)evbuffer_pullup(buf, size);
        if (data)
            reply->body = std::string(data, size);
        evbuffer_drain(buf, size);
    }
}

static void http_error_cb(enum evhttp_request_error err, void *ctx)
{
    HTTPReply *reply = static_cast<HTTPReply*>(ctx);
    reply->error = err;
}

UniValue CallRPC(const std::string& strMethod, const UniValue& params)
{
    std::string host = gArgs.GetArg("-rpcconnect", DEFAULT_RPCCONNECT);
    int port = gArgs.GetArg("-rpcport", BaseParams().RPCPort());

    // Obtain event base
    raii_event_base base = obtain_event_base();

    // Synchronously look up hostname
    raii_evhttp_connection evcon = obtain_evhttp_connection_base(base.get(), host, port);
    evhttp_connection_set_timeout(evcon.get(), gArgs.GetArg("-rpcclienttimeout", DEFAULT_HTTP_CLIENT_TIMEOUT));

    HTTPReply response;
    raii_evhttp_request req = obtain_evhttp_request(http_request_done, (void*)&response);
    if (req == NULL)
        throw std::runtime_error("create http request failed");
    evhttp_request_set_error_cb(req.get(), http_error_cb);

    // Get credentials
    std::string strRPCUserColonPass;
    if (gArgs.GetArg("-rpcpassword", "") == "") {
        // Try fall back to cookie-based authentication if no password is provided
        if (!GetAuthCookie(&strRPCUserColonPass)) {
            throw std::runtime_error(strprintf(
                 _("Could not locate RPC credentials. No authentication cookie could be found, and no rpcpassword is set in the configuration file (%s)"),
                    GetConfigFile(gArgs.GetArg("-conf", MUBDI_CONF_FILENAME)).string().c_str()));

        }
    } else {
        strRPCUserColonPass = gArgs.GetArg("-rpcuser", "") + ":" + gArgs.GetArg("-rpcpassword", "");
    }

    struct evkeyvalq* output_headers = evhttp_request_get_output_headers(req.get());
    assert(output_headers);
    evhttp_add_header(output_headers, "Host", host.c_str());
    evhttp_add_header(output_headers, "Connection", "close");
    evhttp_add_header(output_headers, "Authorization", (std::string("Basic ") + EncodeBase64(strRPCUserColonPass)).c_str());

    // Attach request data
    std::string strRequest = JSONRPCRequestObj(strMethod, params, 1).write() + "\n";
    struct evbuffer* output_buffer = evhttp_request_get_output_buffer(req.get());
    assert(output_buffer);
    evbuffer_add(output_buffer, strRequest.data(), strRequest.size());

    // check if we should use a special wallet endpoint
    std::string endpoint = "/";
    if (!gArgs.GetArgs("-rpcwallet").empty()) {
        std::string walletName = gArgs.GetArg("-rpcwallet", "");
        char* encodedURI = evhttp_uriencode(walletName.c_str(), walletName.size(), false);
        if (encodedURI) {
            endpoint = "/wallet/"+ std::string(encodedURI);
            free(encodedURI);
        } else {
            throw CConnectionFailed("uri-encode failed");
        }
    }
    int r = evhttp_make_request(evcon.get(), req.get(), EVHTTP_REQ_POST, "/");
    req.release(); // ownership moved to evcon in above call
    if (r != 0) {
        throw CConnectionFailed("send http request failed");
    }

    event_base_dispatch(base.get());

    if (response.status == 0)
        throw CConnectionFailed(strprintf("couldn't connect to server: %s (code %d)\n(make sure server is running and you are connecting to the correct RPC port)", http_errorstring(response.error), response.error));
    else if (response.status == HTTP_UNAUTHORIZED)
        throw std::runtime_error("incorrect rpcuser or rpcpassword (authorization failed)");
    else if (response.status >= 400 && response.status != HTTP_BAD_REQUEST && response.status != HTTP_NOT_FOUND && response.status != HTTP_INTERNAL_SERVER_ERROR)
        throw std::runtime_error(strprintf("server returned HTTP error %d", response.status));
    else if (response.body.empty())
        throw std::runtime_error("no response from server");

    // Parse reply
    UniValue valReply(UniValue::VSTR);
    if (!valReply.read(response.body))
        throw std::runtime_error("couldn't parse reply from server");
    const UniValue& reply = valReply.get_obj();
    if (reply.empty())
        throw std::runtime_error("expected reply to have result, error and id properties");

    return reply;
}

int CommandLineRPC(int argc, char* argv[])
{
    std::string strPrint;
    int nRet = 0;
    try {
        // Skip switches
        while (argc > 1 && IsSwitchChar(argv[1][0])) {
            argc--;
            argv++;
        }

        // Method
        if (argc < 2)
            throw std::runtime_error("too few parameters");
        std::string strMethod = argv[1];

        // Parameters default to strings
        std::vector<std::string> strParams(&argv[2], &argv[argc]);
        UniValue params;
        if(gArgs.GetBoolArg("-named", DEFAULT_NAMED)) {
            params = RPCConvertNamedValues(strMethod, strParams);
        } else {
            params = RPCConvertValues(strMethod, strParams);
        }

        // Execute and handle connection failures with -rpcwait
        const bool fWait = gArgs.GetBoolArg("-rpcwait", false);
        do {
            try {
                const UniValue reply = CallRPC(strMethod, params);

                // Parse reply
                const UniValue& result = find_value(reply, "result");
                const UniValue& error = find_value(reply, "error");

                if (!error.isNull()) {
                    // Error
                    int code = error["code"].get_int();
                    if (fWait && code == RPC_IN_WARMUP)
                        throw CConnectionFailed("server in warmup");
                    strPrint = "error: " + error.write();
                    nRet = abs(code);
                    if (error.isObject()) {
                        UniValue errCode = find_value(error, "code");
                        UniValue errMsg  = find_value(error, "message");
                        strPrint = errCode.isNull() ? "" : "error code: "+errCode.getValStr()+"\n";

                        if (errMsg.isStr())
                            strPrint += "error message:\n"+errMsg.get_str();

                        if (errCode.isNum() && errCode.get_int() == RPC_WALLET_NOT_SPECIFIED) {
                            strPrint += "\nTry adding \"-rpcwallet=<filename>\" option to pivx-cli command line.";
                        }
                    }
                } else {
                    // Result
                    if (result.isNull())
                        strPrint = "";
                    else if (result.isStr())
                        strPrint = result.get_str();
                    else
                        strPrint = result.write(2);
                }
                // Connection succeeded, no need to retry.
                break;
            } catch (const CConnectionFailed& e) {
                if (fWait)
                    MilliSleep(1000);
                else
                    throw;
            }
        } while (fWait);
    } catch (const boost::thread_interrupted&) {
        throw;
    } catch (const std::exception& e) {
        strPrint = std::string("error: ") + e.what();
        nRet = EXIT_FAILURE;
    } catch (...) {
        PrintExceptionContinue(NULL, "CommandLineRPC()");
        throw;
    }

    if (strPrint != "") {
        fprintf((nRet == 0 ? stdout : stderr), "%s\n", strPrint.c_str());
    }
    return nRet;
}

int main(int argc, char* argv[])
{
#ifdef WIN32
    util::WinCmdLineArgs winArgs;
    std::tie(argc, argv) = winArgs.get();
#endif
    SetupEnvironment();
    if (!SetupNetworking()) {
        fprintf(stderr, "Error: Initializing networking failed\n");
        return EXIT_FAILURE;
    }
    event_set_log_callback(&libevent_log_cb);

    try {
        if (!AppInitRPC(argc, argv))
            return EXIT_FAILURE;
    } catch (const std::exception& e) {
        PrintExceptionContinue(&e, "AppInitRPC()");
        return EXIT_FAILURE;
    } catch (...) {
        PrintExceptionContinue(NULL, "AppInitRPC()");
        return EXIT_FAILURE;
    }

    int ret = EXIT_FAILURE;
    try {
        ret = CommandLineRPC(argc, argv);
    } catch (const std::exception& e) {
        PrintExceptionContinue(&e, "CommandLineRPC()");
    } catch (...) {
        PrintExceptionContinue(NULL, "CommandLineRPC()");
    }
    return ret;
}
